documentation/API-design-guidelines.md

```diff
@@ -1116,17 +1101,73 @@ When IpAddr is used in a payload the property objectType MUST be present to indi
+
+* `resource` is optional. For APIs with several `paths`, it may include the resource in the path. For example, from `/qod/v0/sessions/{sessionId}`, it would be `sessions`.
 
-Finally, this part describes the OAuth security applied to the API. This spec is for client testing purposes only, but
-there should be as similar as possible to the OAuth flows in your production environment. This definition has the
-following aspects:
+* `action`: There are two cases:
+  - For POST operations with a verb, it will be the verb. For example, from `POST /location-verification/v0/verify`, it would be `verify`.
+    - For endpoints designed as POST but with underlying logic retrieving information, a CRUD action `read` may be added, but if it is a path with single operation and it is not expected to have more operations on it, the CRUD action is not necessary.
+  - For CRUD operations on a resource in paths, it will be one of:
+    - `create`: For operations creating the resource, typically `POST`.
+    - `read`: For operations accessing to details of the resource, typically `GET`.
+    - `update`: For operations modifying the resource, typically `PUT` or `PATCH`.
+    - `delete`: For operations removing the resource, typically `DELETE`.
+    - `write` : For operations creating or modifying the resource, when differentiation between `create` and `update` is not needed.
+
+* Eventually we may need to add an additional level to the scope, such as `api-name:[resource:]action[:detail]`, to deal with cases when only a set of parameters/information has to be allowed to be returned. Guidelines should be enhanced when those cases happen.
+
+##### Examples
+
+| API | path | method | scope |
+| --- | --- | --- | --- |
+| location-verification | /verify | POST | `location-verification:verify` |
+| qod | /sessions | POST | `qod:sessions:create`, or<br>`qod:sessions:write` |
+| qod | /qos-profiles | GET | `qod:qos-profiles:read` |
+
+
+#### 11.6.2 Use of openIdConnect for `securitySchemes`
+
+Generally, OpenID Connect is the protocol to be used for securitization. Each API spec will define the following entry in `securitySchemes`, as shown in [CAMARA-AuthN-AuthZ-Concept.md](https://github.com/camaraproject/IdentityAndConsentManagement/tree/main/documentation/CAMARA-AuthN-AuthZ-Concept.md) document:
+
+```
+components:
+  securitySchemes:
+    openId:
+      type: openIdConnect
+      openIdConnectUrl: https://example.com/.well-known/openid-configuration
+```
+
+The value for `openIdConnectUrl` in the CAMARA spec is an example, that must be substituted by the specific discovery endpoint for OIDC protocol of the API provider, when the API is exposed in one of its environments.
+
+#### 11.6.3 Use of `security` property
+
+Generally, each operation will be protected by a scope and it will include a `security` property with a single element in the array:
```

Comment:
I don't think "Generally" is required here. The rule should be that every operation should be protected by a scope. If there are exceptions, they should be explicitly listed.

An example for client credentials should also be given.

Reply:
@eric-murray I will formulate this way "As soon as user data are manipulated by an operation, the operation must be protected by a scope".

Reply:
So the proposal is to use `openIdConnect` but with no scope when user data is not involved? Something like:
```
  security:
  - openid: []
```

I would clarify that and formulate it the other way around:
"The operation must be protected by a scope, unless no user data will be manipulated by the operation in which case no scope need be specified for the `openIdConnect` security scheme."

An example of specifying no scope might also be useful.

Reply:
I agree that "Generally" must be avoid and give precise guidelines. In OAS we can define a security entry at API root level, which applies to all operations which do not overwrite security. In this sense, it will be better to add what @eric-murray  suggests to a generic root section. Or we may agree that all operations must be protected by some scope, which I think is a good practice and what is commonly happening already in all of our specs.

Reply:
@jlurien @eric-murray I'm fine with Eric proposal: "_The operation must be protected by a scope, unless no user data will be manipulated by the operation in which case no scope need be specified for the openIdConnect security scheme._" 

If we provide for example API of UE density on a geo zone, we haven't user data here and so we did not need a scope - correct?

Reply:
@bigludo7
My concern with allowing "no scope" is that we might think there is no user data involved, but later find that privacy / legal teams say there is. Always having a scope for each operation, even if we think granting this scope will be automatic, allows for additional checks to be added later. I don't see that this complicates things particularly, but I&C sub-project would know better than me.

As an example, the use case you mention (UE density) is coming via GSMA, and is one Vodafone have been looking at. The view of our privacy teams was that this data was "analytics", and hence any subscribers included in the count would have to have opted-in for that.

Now, of course, in this case it would be ridiculous to check which subscribers had consented and calculate the UE density dynamically on that basis for every request for UE density information. The data would just be updated periodically taking this consent into account, but could then be shared freely. However, I give this example to show that it is not always obvious whether an operation is "doing something" with user data that requires consent.

Reply:
Agree also, I think it is good practice to add scopes to all operations. As legal and regulatory considerations are complex and difficult to predict, it is safer to avoid making assumptions while defining the APIs

Reply:
@eric-murray , always having a scope sounds good with an option to grant it automatically sounds like a good way forward.

I suggest to avoid "user data are manipulated by an operation" - this allows different interpretations. One can believe that "manipulation" means "changing". then reading user's location is not a "manipulation".

Reply:
This section is duplicating the work in I&CM, so I will reference the work there